Trigger scintillator hit clustering and tracking

bryngemark opened the pull request:
A rather simple (given the simple geometry) clustering and (straight-line) tracking algorithm for the trigger scintillator electron counting. Would love to have this in the upcoming release.

It is based on the logic that any straight path beam electron should pass through two bars, adjacent in channel number: one per front/back layers of a pad, and that its y coordinate (channel number) should be close to constant (bending is in x). 

And a fix for the duplicated noise hit IDs. 

Based on the recently merged det ID fix branch (#879 ). Related issues: closes #755 , closes #756 , and closes #760 .

awhitbeck commented:
does this include Jeremy's detID fixes?  Is there a separate PR for that which is still pending?

bryngemark commented:
It does. I added this stuff to the branch I used for validation of his changes. The algorithms are independent of that but the duplicate noise ID fix was added to the digi code as it stands after his fixes (the actual fix is one line of code though, and again doesn't explicitly depend on his ID fixes).

bryngemark commented:
Or wait that's not what you asked – you were asking if _this_ is where Jeremy's changes get merged. I believe they already have been. The commit history is from that branch though since I pushed a fix to the DQM code.

omar-moreno commented:
Yes, Jeremy's changes are already in master.

On Sun, Aug 16, 2020, 5:12 PM bryngemark <notifications@github.com> wrote:

> Or wait that's not what you asked – you were asking if *this* is where
> Jeremy's changes get merged. I believe they already have been. The commit
> history is from that branch though since I pushed a fix to the DQM code.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/LDMX-Software/ldmx-sw/pull/884#issuecomment-674597587>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4JMXDVVKEENPF25NWWCG3SBBYXZANCNFSM4P67Z2BQ>
> .
>

tomeichlersmith commented:
@bryngemark Could you rebase your changes onto the current master? That will help us see what changes you are actually making. _Warning:_ rebasing is difficult so be careful and you may need to re-do some patches/commits that occurred before you merged in Jeremy's ID changes.

bryngemark commented:
ok, should be done, easier now?

tomeichlersmith commented:
Yes, much better, thank you!

tomeichlersmith commented on Event/include/Event/TrigScintCluster.h:
```diff
@@ -0,0 +1,191 @@
+#include "Event/EcalCluster.h"
+// these two can't be found..?
+//#include "Framework/ParameterSet.h" // Needed to import parameters from configuration file
+//#include "Framework/EventProcessor.h"
```
You can delete these. The Event module _should not_ depend on the Framework module, so if it compiles without these, you can skip them.

tomeichlersmith commented on Event/include/Event/TrigScintCluster.h:
```diff
@@ -0,0 +1,191 @@
+// ROOT
+#include "TObject.h"
+#include "TString.h"
+#include "TRefArray.h"
```
Please remove TString.h and TRefArray.h includes. I don't think you use them here.

tomeichlersmith commented on EventProc/python/trigScintClustering.py:
```diff
@@ -0,0 +1,50 @@
+        super().__init__(name,'ldmx::TrigScintClusterProducer')
+
+        from LDMX.EventProc import include
+        include.library()
```
In order to update the container with the cmake changes, I removed the automaic generation of the `include.py` module. I replaced it with a change to the constructor:
```
super().__init__(name,'ldmx::TrigScintClusterProducer','EventProc')
```

bryngemark commented on EventProc/src/TrigScintTrackProducer.cxx:
```diff
@@ -0,0 +1,365 @@
+	  
+	  if (verbose_ )
+		{
+		  std::cout << "In TrigScintTrackProducer: Closing file!" << std::endl;
```
I've done this all the way up to the Print() method; I don't want to hardwire its verbosity in its definition. Any good suggestions on how to redirect it to the logger?

bryngemark commented on Event/include/Event/TrigScintCluster.h:
```diff
@@ -0,0 +1,175 @@
+   * @class TrigScintCluster 
+   * @brief Stores cluster information from the trigger scintillator pads. Adds on the ECal cluster functionality
+   */
+  class TrigScintCluster : public EcalCluster {
```
Done now, inherits from TObject instead.

bryngemark commented:
I've done as much as I plan to do here; indentation fixes sound simple but is, I think, a matter of emacs vs vi and not as trivial as it looks at first glance. (I have tried different tab setting solutions and none really works, and honestly, I'm perfectly content with the way emacs normally and beautifully handles this for me and don't want to spend more time on chasing a solution.)

jmmans commented:
Dear all,

I agree.  I think if LDMX is going to use a very unusual set of tabbing 
rules, the emacs bindings should be provided.  The LDMX rules are 
"close" to the cc-mode in emacs, but LDMX for some reason tabs in things 
like "public:".

Jeremy

On 9/3/20 6:13 PM, bryngemark wrote:
> I've done as much as I plan to do here; indentation fixes sound simple 
> but is, I think, a matter of emacs vs vi and not as trivial as it looks 
> at first glance. (I have tried different tab setting solutions and none 
> really works, and honestly, I'm perfectly content with the way emacs 
> normally and beautifully handles this for me and don't want to spend 
> more time on chasing a solution.)
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/LDMX-Software/ldmx-sw/pull/884#issuecomment-686810235>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABFUPYFBYH5IAULCVJ72OK3SEAPJ7ANCNFSM4P67Z2BQ>.
>

omar-moreno commented:
Before doing this, we need to decide on a style and stick with it.  At the
moment, we don't have a set style other than a few suggestions on the wiki
and 4 space indents.

Given that there isn't a standard style, I suggest we either consider
Bjarne's

https://www.stroustrup.com/bs_faq2.html

Or Google's style

https://google.github.io/styleguide/cppguide.html?cl=282453546

I'm biased towards Google's style since I tend to use it on personal
projects and find it nice. However, both are good and well documented.

Once we decide on a style, I can go through and refactor the code and put
in rules that enforces the style before a PR is merged. We can then provide
whatever bindings people might need to get the style into their editors.


On Thu, Sep 3, 2020, 6:05 PM Jeremiah Mans <notifications@github.com> wrote:

> Dear all,
>
> I agree. I think if LDMX is going to use a very unusual set of tabbing
> rules, the emacs bindings should be provided. The LDMX rules are
> "close" to the cc-mode in emacs, but LDMX for some reason tabs in things
> like "public:".
>
> Jeremy
>
> On 9/3/20 6:13 PM, bryngemark wrote:
> > I've done as much as I plan to do here; indentation fixes sound simple
> > but is, I think, a matter of emacs vs vi and not as trivial as it looks
> > at first glance. (I have tried different tab setting solutions and none
> > really works, and honestly, I'm perfectly content with the way emacs
> > normally and beautifully handles this for me and don't want to spend
> > more time on chasing a solution.)
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/LDMX-Software/ldmx-sw/pull/884#issuecomment-686810235>,
>
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/ABFUPYFBYH5IAULCVJ72OK3SEAPJ7ANCNFSM4P67Z2BQ
> >.
> >
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/LDMX-Software/ldmx-sw/pull/884#issuecomment-686840433>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4JMXDSN6HCKCIROPMJJWDSEA4NZANCNFSM4P67Z2BQ>
> .
>

bryngemark commented:
So -- where are we on this? I just pushed a tiny tweak, and I see some test failed: Ecal PN simulation, but target PN etc seems to have run just fine. Whose responsibility is it to make my branch pass the tests and also, against what is it tested..? I'm pretty sure none of the automatic (failing) tests relate to my development at all.

tomeichlersmith commented:
I have updated some of the tests and so it might be a good time to re-merge (or re-rebase) this branch onto master. This should fix up the tests that aren't related to your work. Feel free to slack me if you run into weird git issues.

omar-moreno commented:
Please merge in master.

omar-moreno commented on Event/include/Event/TrigScintCluster.h:
```diff
@@ -0,0 +1,175 @@
+    
+    void setSeed( int idx ) { seed_ = idx ; } 
+	
+	void setEnergy(double energy) {
```
Can you fix the formatting?

omar-moreno commented on Event/include/Event/TrigScintCluster.h:
```diff
@@ -0,0 +1,235 @@
+   * @class TrigScintCluster 
+   * @brief Stores cluster information from the trigger scintillator pads. Adds on the ECal cluster functionality
+   */
+  class TrigScintCluster : public TObject {
```
Inheriting from TObject is not necessary.

omar-moreno commented on Event/src/TrigScintCluster.cxx:
```diff
@@ -0,0 +1,41 @@
+
+namespace ldmx {
+
+  TrigScintCluster::TrigScintCluster() { //: EcalCluster() {
```
Since you are using the default cluster, there is no need to have an empty constructor in the src file.  Instead, in the header file just do 
    `TrigScintCluster() = default;`

omar-moreno commented on Event/src/TrigScintCluster.cxx:
```diff
@@ -0,0 +1,41 @@
+  
+  TrigScintCluster::~TrigScintCluster() {
+    Clear();
+    
```
Fix extra space.

omar-moreno commented on Event/src/TrigScintCluster.cxx:
```diff
@@ -0,0 +1,41 @@
+	centroidZ_ = 0;
+	setEnergy(0);
+	setNHits(0);
+    setCentroid(0); 
```
Fix formatting.

omar-moreno commented:
@bryngemark Can you fix the formatting of all the code?

omar-moreno commented on EventProc/src/TrigScintTrackProducer.cxx:
```diff
@@ -0,0 +1,365 @@
+	  // TO DO: allow any number of input collections
+    
+
+	  if (verbose_)
```
Why do you need a check for verbose if you are already setting the log level to info? You can always change the log level to something higher if you want these messages to stop.

bryngemark replied:
So to my understanding, the logger verbosity level is a global feature (so if I set it to debug, I get debug level messages from all invoked producers). For my use of clustering and tracking, which is still to work out the best parameters, it's very convenient to be able to just make one of the producers verbose. So I'd like to keep this configurable per producer.

omar-moreno replied:
You bring up a good point.  We should be able to set the log level at the processor level. Totally fine to leave as is then.

omar-moreno commented on EventProc/src/TrigScintTrackProducer.cxx:
```diff
@@ -0,0 +1,365 @@
+	  }
+    
+    
+	  const auto seeds{event.getCollection< TrigScintCluster >(seeding_collection_, passName_)}; 
```
Should use the RandomNumberSeedService for this instead.

bryngemark commented:
Ok.

I've fixed the indentation issues as much as I could. Again it doesn't look the same for me (in emacs) as on github, and I don't want to spend any more time chasing whitespace. I look forward to a common solution like the emacs bindings Jeremy suggested. 

As for the failed tests, which are more recent than my changes = independent of this work, I have attempted rebasing, but the changes (to other modules etc) are substantial, and it fails. I do think this is completely unnecessary for merging in the PR. The tests don't test this code anyway.

The two conflicting files are:
1. `trigScintDigis.py` (governing hit reconstruction) --> this I have replaced by a common python configuration for hit reconstruction, clustering and tracking: `trigScint.py`, so this "conflict" is actually a (to the user, anyway disruptive) file deletion
2. `TrigScintDigiProducer.cxx` --> here all I did was add one line of bugfix, to record the noise hit ID. It _should_ be an easy merge?

awhitbeck commented on Event/include/Event/TrigScintCluster.h:
```diff
@@ -0,0 +1,230 @@
+   * @class TrigScintCluster 
+   * @brief Stores cluster information from the trigger scintillator pads. Adds on the ECal cluster functionality
+   */
+  class TrigScintCluster : {
```
I think this is a typo.  I guess we're not intending this to inherit from anything.  The code also won't compile like this.

omar-moreno commented:
Thanks for all of the updates, I'll do some final formatting, make sure it compiles, merge in master and merge it in.

omar-moreno commented:
> I have updated some of the tests and so it might be a good time to re-merge (or re-rebase) this branch onto master. This should fix up the tests that aren't related to your work. Feel free to slack me if you run into weird git issues.
> (That being said, @omar-moreno I think this is good and could be merged even though the tests are failing)

@tomeichlersmith where the changes you made pushed to Framework? The issue with the test is still present so was wondering if I needed to update a submodule.

tomeichlersmith commented:
Yes, there were changes in Framework, so the submodules need to be updated.

On Mon, Oct 12, 2020 at 5:26 PM Omar Moreno <notifications@github.com>
wrote:

> I have updated some of the tests and so it might be a good time to
> re-merge (or re-rebase) this branch onto master. This should fix up the
> tests that aren't related to your work. Feel free to slack me if you run
> into weird git issues.
> (That being said, @omar-moreno <https://github.com/omar-moreno> I think
> this is good and could be merged even though the tests are failing)
>
> @tomeichlersmith <https://github.com/tomeichlersmith> where the changes
> you made pushed to Framework? The issue with the test is still present so
> was wondering if I needed to update a submodule.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LDMX-Software/ldmx-sw/pull/884#issuecomment-707375836>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHT5H7VKMTN7HAROQLOXQ4TSKN7ALANCNFSM4P67Z2BQ>
> .
>

tomeichlersmith commented:
@bryngemark I figured out why the tests are failing! I don't know why the log from the action doesn't show this, but when I ran the test configs on my local machine I got this:
```
---- LDMXSW: Loading configuration --------
Traceback (most recent call last):
  File "/home/tom/ldmx/ldmx-sw/EventProc/test/sim_and_reco.py", line 28, in <module>
    from LDMX.EventProc.trigScintDigis import TrigScintDigiProducer
ModuleNotFoundError: No module named 'LDMX.EventProc.trigScintDigis'
```
It looks like you need to update `EventProc/test/sim_and_reco.py` to match the new module name that you use. (My guess is that the merging of master changed your changes back).

bryngemark commented:
Thanks for figuring this out, @tomeichlersmith! I have updated the name of the trig scint reco config in the test. I also added in clustering and tracking to the test sequence.

bryngemark commented:
It does look, though, that the change in how the random seeds are generated was not fully propagated here (tests fail). @tomeichlersmith, @omar-moreno, I'm guessing one of you knows how to make the TS digi code compliant with the new approach?

tomeichlersmith commented:
Yea, the below code snippet is from EcalDigiProducer and shows how to seed the noise generator while using the RNSS.
https://github.com/LDMX-Software/Ecal/blob/22b54afbe6b4e220508a957a3dae717e9aa48b58/src/EcalDigiProducer.cxx#L55-L59

bryngemark commented:
Thanks @tomeichlersmith! I added that, and just ran the test locally and it didn't fail -- so, fingers crossed then.
